src/ivc/step_folding_circuit.rs

```diff
@@ -0,0 +1,401 @@
+            config.instance,
+            0,
+        )?;
+        layouter.constrain_instance(output_hash.cell(), config.instance, 1)?;
+
+        Ok(z_output)
+    }
+}
+
+impl<'link, const ARITY: usize, C, SC, RO, const T: usize>
+    StepFoldingCircuit<'link, ARITY, C, SC, RO, T>
+where
+    C: CurveAffine,
+    C::Base: ff::PrimeFieldBits + ff::FromUniformBytes<64>,
+    C::Scalar: ff::PrimeFieldBits + ff::FromUniformBytes<64>,
+    SC: StepCircuit<ARITY, C::Base> + Sized,
+    RO: ROCircuitTrait<C::Base, Config = MainGateConfig<T>>,
+{
+    pub fn synthesize(
+        &self,
+        config: StepConfig<ARITY, C::Base, SC, T>,
+        layouter: &mut impl Layouter<C::Base>,
+    ) -> Result<[C::Base; ARITY], SynthesisError> {
+        let assigned_z_i: [_; ARITY] = layouter.assign_region(
+            || "assigned_zi",
+            |region| {
+                let mut region = RegionCtx::new(region, 0);
+
+                config
+                    .main_gate_config
+                    .advice_cycle_assigner()
+                    .assign_all_advice(&mut region, || "z0_primary", self.input.z_0.iter().copied())
+                    .map(|inp| inp.try_into().unwrap())
+            },
+        )?;
+
+        let assigned_z_next = <Self as StepCircuit<ARITY, C::Base>>::synthesize_step(
+            self,
+            config,
+            layouter,
+            &assigned_z_i,
+        )?;
+
+        Ok(assigned_z_next.map(|cell| cell.value().unwrap().cloned().unwrap()))
+    }
+
+    fn synthesize_step_base_case(
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
+        public_params: &StepParams<C::Base, RO>,
+        u: &PlonkInstance<C>,
+        config: MainGateConfig<T>,
+    ) -> Result<AssignedRelaxedPlonkInstance<C>, SynthesisError> {
+        let Unew_base = layouter.assign_region(
+            || "synthesize_step_base_case",
+            move |region| {
+                let chip = if public_params.is_primary_circuit {
+                    FoldRelaxedPlonkInstanceChip::new_default(
+                        public_params.limb_width,
+                        public_params.limbs_count,
+                        u.challenges.len(),
+                        u.W_commitments.len(),
+                        config.clone(),
+                    )
+                } else {
+                    FoldRelaxedPlonkInstanceChip::from_instance(
+                        u.clone(),
+                        public_params.limb_width,
+                        public_params.limbs_count,
+                        config.clone(),
+                    )
+                };
+
+                Ok(chip.assign_current_relaxed(&mut RegionCtx::new(region, 0))?)
+            },
+        )?;
+
+        Ok(Unew_base)
+    }
+
+    fn synthesize_step_non_base_case(
+        &self,
+        config: &StepConfig<ARITY, C::Base, SC, T>,
+        layouter: &mut impl Layouter<C::Base>,
+        input: &StepInputs<ARITY, C, RO>,
```

Comment:
instead of `StepInputs` here, use `AssignedStepInputs`

Reply:
https://github.com/snarkify/sirius/issues/153